src/ddsrt/include/dds/ddsrt/events.h

```diff
@@ -0,0 +1,208 @@
+  /**< timeout has occurred on the resource */
+//#define DDSRT_EVENT_FLAG_TIMEOUT (1u<<4)
+  /**< notification of ip address change on the resource*/
+//#define DDSRT_EVENT_FLAG_IP_CHANGE (1u<<5)  
+///@}
+
+  /**
+  * @brief Describes the type of object being monitored for events.
+  */
+ typedef enum ddsrt_event_type {
+    DDSRT_EVENT_TYPE_UNSET, /**< uninitialized state*/
+    DDSRT_EVENT_TYPE_SOCKET /**< indicating a socket type connection*/
+    /*,
+    DDSRT_EVENT_TYPE_FILE,
+    DDSRT_EVENT_TYPE_IFADDR*/
+  } ddsrt_event_type_t;
+
+  /**
+  * @brief Structure containing the information of which object is being watched.
+  *
+  * Is used by ddsrt_event_queue to keep track of objects being watched.
+  */
+  struct ddsrt_event {
+    uint32_t flags; /**< Type of event being watched for, composited from DDSRT_EVENT_FLAG_$EVENT_TYPE$.*/
+    ddsrt_event_type_t type; /**< Type of object being watched.*/
+    ddsrt_atomic_uint32_t triggered; /**< Trigger status after a call to ddsrt_event_queue_wait.*/
+    union {
+      struct { ddsrt_socket_t sock; } socket;
+      /*struct ddsrt_event_file file;  this is for future expansion to be able to register events on files*/
+      /*struct ddsrt_event_ifaddr ifaddr; this is for future expansion to be able to register IP address changes*/
+    } data; /**< Container for the object being watched.*/
+  };
+
+  typedef struct ddsrt_event ddsrt_event_t;
+
+  typedef struct ddsrt_event_queue ddsrt_event_queue_t;
+  
+  /**
+  * @brief Initializes an existing event to indicate a socket.
+  *
+  * This function will set the type to be DDSRT_EVENT_TYPE_SOCKET, the triggered status to DDSRT_EVENT_FLAG_UNSET and
+  * the flags and socket fields to the values supplied.
+  *
+  * @param[in,out] ev Pointer to the event to initialize.
+  * @param[in] sock Socket to initialize the event with.
+  * @param[in] flags Flags to set for the event.
+  *
+  * @retval DDS_RETCODE_OK
+  *             In all cases.
+  */
+  DDS_EXPORT dds_return_t ddsrt_event_socket_init(ddsrt_event_t* ev, ddsrt_socket_t sock, uint32_t flags) ddsrt_nonnull((1));
+
+  /**
+  * @brief Event queue creation function.
+  *
+  * This function will attempt to reserve memory for the event queue and open any necessary additional resources, i.e.:
+  * - a pipe for interrupting its own wait state
+  * - an instance to a kernel event monitor (in the case of BSD operating systems)
+  *
+  * @returns !NULL The event queue that was created.
+  * @returns NULL Event queue could not be created succesfully.
+  */
+  DDS_EXPORT ddsrt_event_queue_t* ddsrt_event_queue_create(void);
+
+  /**
+  * @brief Cleans up an event queue.
+  *
+  * Will close/free any resources managed by the event_queue (if any), then free the memory of the queue.
+  *
+  * @param[in,out] queue The queue to clean up.
+  *
+  * @retval DDS_RETCODE_OK
+  *             In all cases.
+  */
+  DDS_EXPORT dds_return_t ddsrt_event_queue_delete(ddsrt_event_queue_t* queue) ddsrt_nonnull_all;
```

Comment:
If it's always `DDS_RETCODE_OK`, wouldn't it be better to just make it `void`?

Reply:
This was one of the things I was planning on going over in the next iteration.

Reply:
This also touches on a similar question made by @eboasson .

Reply:
This is now implemented.